code/ed25519/Hacl.Impl.SHA512.ModQ.fst

```diff
@@ -1,31 +1,32 @@
 module Hacl.Impl.SHA512.ModQ
 
-module ST = FStar.HyperStack.ST
-open FStar.HyperStack.All
+open FStar.HyperStack
+open FStar.HyperStack.ST
+open FStar.Mul
 
 open Lib.IntTypes
-open Lib.ByteSequence
-open Lib.Sequence
 open Lib.Buffer
 
-module F56 = Hacl.Impl.BignumQ.Mul
+module ST = FStar.HyperStack.ST
+module BSeq = Lib.ByteSequence
+
 module Hash = Hacl.Streaming.SHA2
+module F56 = Hacl.Impl.BignumQ.Mul
 
 #reset-options "--z3rlimit 50 --fuel 0 --ifuel 0"
 
 val sha512_pre_msg:
     hash:lbuffer uint8 64ul
   -> prefix:lbuffer uint8 32ul
-  -> len:size_t{v len + 32 <= max_size_t}
+  -> len:size_t
   -> input:lbuffer uint8 len ->
   Stack unit
     (requires fun h ->
       live h hash /\ live h prefix /\ live h input /\
       disjoint input hash /\ disjoint prefix hash)
     (ensures fun h0 _ h1 -> modifies (loc hash) h0 h1 /\
       as_seq h1 hash == Spec.Agile.Hash.hash Spec.Hash.Definitions.SHA2_512
-           (concat #uint8 #32 #(v len) (as_seq h0 prefix) (as_seq h0 input))
-    )
+        (Seq.append (as_seq h0 prefix) (as_seq h0 input)))
```

Comment:
Is this where the restriction on the extra 64 bytes is lifted? Because you use `Seq.append` which can reason on > uint32_max long sequences...?

Reply:
yes, that's correct